src/System.Web.OData/OData/Query/Expressions/PropertyContainer.cs

```diff
@@ -233,57 +273,108 @@ public override object GetValue()
         }
 
         // Entityframework requires that the two different type initializers for a given type in the same query have the same set of properties in the same order.
-        // A $select=Prop1,Prop2 where Prop1 and Prop2 are of the same type without this extra NamedPropertyWithNext type results in an select expression that looks like,
-        //      c => new NamedProperty<int> { Name = "Prop1", Value = c.Prop1, Next = new NamedProperty<int> { Name = "Prop2", Value = c.Prop2 } };
+        // A $select=Prop1,Prop2,Prop3 where Prop1 and Prop2 are of the same type without this extra NamedPropertyWithNext type results in an select expression that looks like,
+        //      c => new NamedProperty<int> { Name = "Prop1", Value = c.Prop1, LeftNext = new NamedProperty<int> { Name = "Prop2", Value = c.Prop2 }, RightNext = new NamedProperty<int> { Name = "Prop3", Value = c.Prop3 } };
         // Entityframework cannot translate this expression as the first NamedProperty<int> initialization has Next and the second one doesn't. Also, Entityframework cannot 
         // create null's of NamedProperty<T>. So, you cannot generate an expression like new NamedProperty<int> { Next = null }. The exception that EF throws looks like this,
         // "The type 'NamedProperty`1[SystemInt32...]' appears in two structurally incompatible initializations within a single LINQ to Entities query. 
         // A type can be initialized in two places in the same query, but only if the same properties are set in both places and those properties are set in the same order."
-        internal class NamedPropertyWithNext<T> : NamedProperty<T>
+        internal class NamedPropertyWithNextLeftOnly<T> : NamedProperty<T>
+        {
+            public PropertyContainer LeftNext { get; set; }
+
+            public override void ToDictionaryCore(Dictionary<string, object> dictionary, IPropertyMapper propertyMapper,
+                bool includeAutoSelected)
+            {
+                base.ToDictionaryCore(dictionary, propertyMapper, includeAutoSelected);
+                LeftNext.ToDictionaryCore(dictionary, propertyMapper, includeAutoSelected);
+            }
+        }
+
+        internal class NamedPropertyWithNext<T> : NamedPropertyWithNextLeftOnly<T>
+        {
+            public PropertyContainer RightNext { get; set; }
+
+            public override void ToDictionaryCore(Dictionary<string, object> dictionary, IPropertyMapper propertyMapper,
+                bool includeAutoSelected)
+            {
+                base.ToDictionaryCore(dictionary, propertyMapper, includeAutoSelected);
+                if (RightNext != null)
```

Comment:
Isn't NamedPropertyWithNext used in the case when LeftNext != null && RightNext != null already? Maybe leaving this out would provide more diagnostic information, i.e. crash because NamedPropertyWithNext was used with RightNext == null? 

Also wondering why this has a check for RightNext but NamedPropertyWithNextLeftOnly does not have a check for LeftNext.

Reply:
With more columns Left/Right isn't enough to avoid stackoverflow. I'm implementing B-tree (with 32 branches for now)